localization/ar_tag_based_localizer/src/ar_tag_based_localizer_core.cpp

```diff
@@ -0,0 +1,297 @@
+  */
+  tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+  tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(this);
+
+  /*
+    Initialize image transport
+  */
+  it_ = std::make_unique<image_transport::ImageTransport>(shared_from_this());
+
+  /*
+    Subscribers
+  */
+  image_sub_ = it_->subscribe("image", 1, &ArTagBasedLocalizer::image_callback, this);
+  cam_info_sub_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(
+    "camera_info", 1,
+    std::bind(&ArTagBasedLocalizer::cam_info_callback, this, std::placeholders::_1));
+
+  /*
+    Publishers
+  */
+  using geometry_msgs::msg::PoseWithCovarianceStamped;
+  auto qos = rclcpp::QoS(rclcpp::QoSInitialization::from_rmw(rmw_qos_profile_default));
+  qos.reliable();
+  qos.transient_local();
+  image_pub_ = it_->advertise("result", 1);
+  pose_pub_ = this->create_publisher<PoseWithCovarianceStamped>("pose_with_covariance", qos);
+
+  RCLCPP_INFO(this->get_logger(), "Setup of ar_tag_based_localizer node is successful!");
+  return true;
+}
+
+void ArTagBasedLocalizer::image_callback(const sensor_msgs::msg::Image::ConstSharedPtr & msg)
+{
+  if ((image_pub_.getNumSubscribers() == 0) && (pose_pub_->get_subscription_count() == 0)) {
+    RCLCPP_DEBUG(this->get_logger(), "No subscribers, not looking for ArUco markers");
+    return;
+  }
+
+  if (!cam_info_received_) {
+    RCLCPP_DEBUG(this->get_logger(), "No cam_info has been received.");
+    return;
+  }
+
+  builtin_interfaces::msg::Time curr_stamp = msg->header.stamp;
+  cv_bridge::CvImagePtr cv_ptr;
+  try {
+    cv_ptr = cv_bridge::toCvCopy(*msg, sensor_msgs::image_encodings::RGB8);
+  } catch (cv_bridge::Exception & e) {
+    RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
+    return;
+  }
+
+  cv::Mat in_image = cv_ptr->image;
+
+  // detection results will go into "markers"
+  std::vector<aruco::Marker> markers;
+
+  // ok, let's detect
+  detector_.detect(in_image, markers, cam_param_, marker_size_, false);
+
+  // for each marker, draw info and its boundaries in the image
+  for (const aruco::Marker & marker : markers) {
+    tf2::Transform tf_cam_2_marker = right_to_left_ * aruco_marker_2_tf2(marker);
+
+    geometry_msgs::msg::TransformStamped tf_cam_2_marker_stamped;
+    tf2::toMsg(tf_cam_2_marker, tf_cam_2_marker_stamped.transform);
+    tf_cam_2_marker_stamped.header.stamp = curr_stamp;
+    tf_cam_2_marker_stamped.header.frame_id = camera_frame_;
+    tf_cam_2_marker_stamped.child_frame_id = "detected_marker_" + std::to_string(marker.id);
+    tf_broadcaster_->sendTransform(tf_cam_2_marker_stamped);
+
+    geometry_msgs::msg::PoseStamped pose_cam_2_marker;
+    tf2::toMsg(tf_cam_2_marker, pose_cam_2_marker.pose);
+    pose_cam_2_marker.header.stamp = curr_stamp;
+    pose_cam_2_marker.header.frame_id = std::to_string(marker.id);
+    publish_pose_as_base_link(pose_cam_2_marker);
```

Comment:
Yes, by design, when multiple AR tags are captured at the same time, the number of self-positions are published in a very narrow time interval. The reasons why I think there is no problem with this design are:

(1) Each detected AR tag should output similar poses, so there should be no particular problem. It would be better to mix these in EKF rather than considering averaging at this node.

(2) Currently, only AR tags captured within a 6m distance are used for pose calculation, so realistically, multiple AR tags would not be adopted at the same time.

However, in the test rosbag, AR tags are placed far enough apart from each other, and the above considerations have not been precisely verified.

Reply:
Thanks for the elaboration! Make sense to me.